src/lib.rs

```diff
@@ -127,33 +134,165 @@ pub enum Parsing<T, N = ()> {
     NeedMore(N)
 }
 
+/// A buffer type used for implementing `Extension`s.
```

Comment:
This type is not just for extensions but (also) used to represent any payload to send (at least internally), right? My personal naming suggestion would be `Payload`.

Reply:
Internally and for outbound messages yes, but I do not want to mention this fact as part of the API documentation as it should not matter to the reader. As for the name `Payload` ― it may be a tad too universal here as one may wonder why extensions get passed a `Payload` when encoding, but a `BytesMut` when decoding. The name `Storage` emphasises what this type is about, i.e. to unify different ways to represent data, like a copy-on-write pointer. Also, as far as the user is concerned, (inbound) message payloads are actually represented as `data::Incoming` and `data::Data`. Adding `Payload` to the API would be more confusing, no?

Reply:
> [..] it may be a tad too universal here as one may wonder why extensions get passed a Payload when encoding, but a BytesMut when decoding.

Good point.